Check for updates concurrently

FranciscoTGouveia opened the pull request:
This change is related to #3132.

Currently, the `rustup check` command verifies updates for each toolchain sequentially.
This patch modifies that behavior to perform the checks concurrently.

To implement this change, the `futures_util` crate was introduced. I would appreciate feedback on this choice, and whether there might be a preferable alternative for handling concurrency.

My first attempt at adding concurrency caused all output to be printed only after every channel had been checked, which deviated from the current behavior of the command.
To preserve the existing behavior (reporting progress as each toolchain is checked), I tried printing concurrently as well.
However, this led to race conditions, resulting in inconsistent output order.

To address this, I used the `indicatif` crate (cc @djc) to display progress bars while maintaining the correct output order.

This PR will soon be updated to ensure progress bars are correctly written to _stdout_ in order to pass the tests.

Regarding performance, using `hyperfine`, a **2.2× speedup** was observed compared to the sequential version. These benchmarks were run 100 times (with 5 warmup runs) over a 50 Mbps connection.

FranciscoTGouveia commented:
After taking a closer look at the code, I realized that `buffered` can be replaced with `buffer_unordered` when running in a _tty_, since `indicatif` already ensures that output appears in the correct order.

To see if this had a practical impact, I re-ran the benchmarks (using the same settings as before) and observed a **3.1x** speedup.

rami3l reviewed (approved):
Nice work overall 🎉 I have to suggest some other changes regarding some small details but modulo those everything seems to be going pretty well. Looking forward to landing this patch!

rami3l commented on src/cli/rustup_mode.rs:
```diff
@@ -812,7 +812,7 @@ async fn check_updates(cfg: &Cfg<'_>, opts: CheckOpts) -> Result<utils::ExitCode
                         .tick_chars("⠁⠂⠄⡀⢀⠠⠐⠈ "),
                 );
-                pb.set_message(format!("{} - Checking...", name));
+                pb.set_message(format!("{name} - Checking..."));
```
These clippy lints seem to largely originate from your earlier modifications in this PR; in this case it is recommended to squash them back to the respective original commit(s). You can use the [`git absorb`](https://andrewlock.net/super-charging-git-rebase-with-git-absorb/) plugin to do this more easily if you are using git.

rami3l commented on src/cli/rustup_mode.rs:
```diff
@@ -795,6 +795,11 @@ async fn default_(
 async fn check_updates(cfg: &Cfg<'_>, opts: CheckOpts) -> Result<utils::ExitCode> {
     let t = cfg.process.stdout().terminal(cfg.process);
     let is_a_tty = t.is_a_tty();
+    let use_colors = match cfg.process.var("RUSTUP_TERM_COLOR").as_deref() {
```
Do you think it's useful to expose the `ColorChoice` we have already calculated in `ColorableTerminal::new()` here like what we have done in `.is_a_tty()`, instead of repeating the fetching of `RUSTUP_TERM_COLOR` all over the place? I'd expect that to be reused in many other places in the future so...

FranciscoTGouveia replied:
Yes, I had the same thought. Fetching the `RUSTUP_TERM_COLOR` environment variable in multiple places could definitely hurt readability. I was just a bit hesitant to add yet another method to `ColorableTerminal`, but I completely agree with your suggestion. Thank you!

rami3l commented on src/cli/rustup_mode.rs:
```diff
@@ -824,28 +825,38 @@ async fn check_updates(cfg: &Cfg<'_>, opts: CheckOpts) -> Result<utils::ExitCode
             },
         );
 
+        // If we are running in a TTY, we can use `buffer_unordered` since
```
I suspect this PR can be entirely or partially squashed back to one or more of the previous commits following the same principle... This comment for example feels a bit off in this very commit.

rami3l commented on src/cli/rustup_mode.rs:
```diff
@@ -792,12 +792,17 @@ async fn default_(
     if num_channels > 0 {
-        let mp = MultiProgress::with_draw_target(ProgressDrawTarget::stdout());
-
+        let mp = if is_a_tty {
```
Bonus: `cargo` uses the [`CARGO_TERM_PROGRESS_WHEN`](https://doc.rust-lang.org/cargo/reference/config.html#termprogresswhen) environment variable to control the drawing of the progress bar. I think it'd be interesting to mirror that here as well, maybe in a separate commit or PR.

rami3l commented on src/process/terminalsource.rs:
```diff
@@ -223,6 +232,81 @@ impl io::Write for ColorableTerminalLocked {
 
+impl TermLike for ColorableTerminal {
+    fn width(&self) -> u16 {
+        Term::stdout().size().1
```
Bonus: `cargo` uses the [`CARGO_TERM_PROGRESS_WIDTH`](https://doc.rust-lang.org/cargo/reference/config.html#termprogresswidth) environment variable to control the terminal width being detected. I think it'd be interesting to mirror that here as well, maybe in a separate commit or PR.

rami3l commented on src/cli/rustup_mode.rs:
```diff
@@ -793,44 +795,67 @@ async fn default_(
-            }
-        }
+    let num_channels = channels.len();
+    if num_channels > 0 {
```
Probably worth putting the link (https://github.com/rust-lang/futures-rs/pull/1194#discussion_r209501774) here in the comments saying that `.buffered(0)` would cause a hang.

rami3l commented on src/cli/rustup_mode.rs:
```diff
@@ -793,44 +795,67 @@ async fn default_(
+                        (msg, style)
+                    }
+                    (Some(cv), None) => {
+                        let msg = format!("{} - Up to date : {}", name, cv);
```
IIRC `clippy` would fail on this commit, I suggest that you do another rebase to make sure?

rami3l commented on src/cli/rustup_mode.rs:
```diff
@@ -797,6 +797,8 @@ async fn default_(
 async fn check_updates(cfg: &Cfg<'_>, opts: CheckOpts) -> Result<utils::ExitCode> {
     let t = cfg.process.stdout().terminal(cfg.process);
     let is_a_tty = t.is_a_tty();
+    let use_colors = t.color_choice() == terminalsource::ColorChoice::Auto
```
Nit: `matches!()` would be a better choice here.

djc replied:
Why? This seems fine to me.

Do suggest importing `ColorChoice` from `terminalsource`.

rami3l replied:
@djc I was thinking about something like `matches!(t.color_choice(), ColorChoice::Auto | ColorChoice::Always)`.

djc replied:
Ahh, yes, that would be better.

djc commented on src/cli/rustup_mode.rs:
```diff
@@ -793,44 +795,67 @@ async fn default_(
-        }
+    let num_channels = channels.len();
+    if num_channels > 0 {
+        let mp = MultiProgress::with_draw_target(ProgressDrawTarget::stdout());
```
I suggest you pick a more explicit variable name than `mp` for this.

djc commented on src/cli/rustup_mode.rs:
```diff
@@ -793,44 +795,67 @@ async fn default_(
+            })
+            .collect();
+
+        let channels = tokio_stream::iter(channels.into_iter().zip(progress_bars))
```
Why are `progress_bars` and `channels` created in separate loops? Suggest just using for-loops here instead of a trivial combinator chain with a long `map()` closure.

rami3l replied:
@djc We are mapping over a stream rather than an iterator, so doing a plain `for` loop might not help here. Depending on the current configuration more combinators will be involved down the line, notably with `.buffered()` and `.buffer_unordered()`.

djc commented on src/cli/rustup_mode.rs:
```diff
@@ -794,12 +794,17 @@ async fn default_(
     if num_channels > 0 {
-        let mp = MultiProgress::with_draw_target(ProgressDrawTarget::stdout());
-
+        let mp = if is_a_tty {
```
Given that this commit ends up modifying a lot of the code introduced in the previous commit, suggest that this should be squashed into the previous one.

djc commented on Cargo.toml:
```diff
@@ -52,8 +52,10 @@ enum-map = "2.5.0"
+futures-util = "0.3.31"
 git-testament = "0.2"
 home = "0.5.4"
+indicatif = "0.17"
```
Note that the indicatif 0.17 version that used console 0.16 was yanked, should use indicatif 0.18 here.

djc commented on src/cli/rustup_mode.rs:
```diff
@@ -810,11 +811,11 @@ async fn check_updates(cfg: &Cfg<'_>, opts: CheckOpts) -> Result<utils::ExitCode
                 let pb = mp.add(ProgressBar::new(1));
                 pb.set_style(
-                    ProgressStyle::with_template("{msg} {spinner:.green}")
+                    ProgressStyle::with_template("{msg:.bold} - Checking... {spinner:.green}")
```
This too is primarily changing code introduced in the previous commit, just squash instead.

djc commented on src/cli/rustup_mode.rs:
```diff
@@ -793,44 +795,67 @@ async fn default_(
+                        .tick_chars("⠁⠂⠄⡀⢀⠠⠐⠈ "),
+                );
+                pb.set_message(format!("{} - Checking...", name));
+                pb.enable_steady_tick(std::time::Duration::from_millis(100));
```
Nit: please import `Duration`.

djc reviewed (commented):
Did you benchmark the performance impact of one or more rustup commands?

Would also be good to add an animation/video of the new output.

FranciscoTGouveia committed:
feat(terminal): implement the `TermLike` trait for `ColorableTerminal`

To later support `indicatif`'s progress bars, this trait was implemented
for the `ColorableTerminal`.

It is important to note that `is_a_tty` and `color_choice` will be
needed for deciding if progress bars are written to the stdout (or
hidden) and if they are styled with colors, respectively.

FranciscoTGouveia committed:
feat(updates): check for updates concurrently

This change implied the introduction of the `futures_util` crate.
Feedback is greately appreciated on the above.

To ensure that there are no hangs, a special "if" was needed to ensure
that we never try running `.buffered(0)`. Otherwise, this would cause an
hang (kudos to rami3l for helping me out with this).

To report progress in a concurrent way whilst maintaining order, the
`indicatf` crate was used.
Note that, if we are not in a *tty* the progress wil be hidden and will
instead be printed after every channels is checked for updates.

NB: since we are using `indicatif` to display progress, when running in
a *tty*, we can use `buffer_unordered` instead of `buffered`, as
`indicatif` already handles the output in the correct order.
A quick benchmark (100 runs + 10 warmpu runs in a 50Mbps connection)
showed that this small change yields a **1.2x** speedup.

FranciscoTGouveia commented:
In order to evaluate the performance impact of these changes, I ran a series of benchmarks using `hyperfine`.
My setup included a 50 Mbps internet connection, 16 GB of RAM, and an i7-1165G7 processor.

I compared my modified version of the `rustup check` command against the current implementation on the master branch.
Each benchmark was executed 100 times, with 5 warmup runs beforehand.
To ensure a meaningful comparison, I checked **20** different toolchains for updates in each run.

The results showed a **3.3×** speedup over the current implementation.
This exceeded my initial expectations, thanks to the realization that, since `indicatif` already handles the output in the correct order, we could safely replace `buffered` with `buffer_unordered`.

To give a better understanding of what this change implies regarding the user experience, I leave below a short animation.
On the left, you’ll see the current (sequential) behavior, and on the right, the new (concurrent) behavior:

![showcase](https://github.com/user-attachments/assets/93cb609e-117c-48d4-bedd-972db05a2a6c)

rami3l commented on src/cli/rustup_mode.rs:
```diff
@@ -792,45 +796,99 @@ async fn default_(
-                let _ = t.reset();
-                writeln!(t.lock(), " : {cv}")?;
+    let num_channels = channels.len();
+    // Ensure that `.buffered()` is never called with 0 as this will cause an hang.
```
Nit: ... a hang.

rami3l commented on src/cli/rustup_mode.rs:
```diff
@@ -792,45 +796,99 @@ async fn default_(
+        };
+
+        let t = cfg.process.stdout().terminal(cfg.process);
+        for result in channels.into_iter() {
```
Nit: `.into_iter()` seems redundant here.

rami3l reviewed (approved):
That's much clearer, thanks!

djc commented:
It seems like a substantial downside to me that the new version takes so long to show any output at all. IMO you should look into fixing that.

rami3l commented:
> It seems like a substantial downside to me that the new version takes so long to show any output at all. IMO you should look into fixing that.

@djc I believe you are reading the gif wrong? This is not a side-by-side comparison; these two commands are executed one after another, both `time`d so that you can see the second is faster.

@FranciscoTGouveia It's true that we would like to eliminate all such confusions in the final demo to be used in the blog post/release notes or whatever it may be.

FranciscoTGouveia commented:
I’m sorry, I should have showcased the implementation a bit better.

In reality, the new version displays output instantaneously. The delay you see between when the left side finishes and the right side starts is simply because I only ran the new version after the base version completed. The pause was caused by me switching panes in `tmux` during the recording.
Below is an updated version of the animation that I hope better illustrates the behavior:

![showcase_final](https://github.com/user-attachments/assets/09084797-f895-4eca-9659-c011031305ae)

djc commented:
Ahh, awesome, that's much more impressive!

djc commented on src/cli/rustup_mode.rs:
```diff
@@ -792,45 +796,99 @@ async fn default_(
+    // Ensure that `.buffered()` is never called with 0 as this will cause a hang.
+    // See: https://github.com/rust-lang/futures-rs/pull/1194#discussion_r209501774
+    if num_channels > 0 {
+        let multi_progress_bars = if is_a_tty {
```
Nit: extract the `MultiProgress::with_draw_target()` out of the `if` (also, I think a `match` would look better than an `if` for this).

rami3l replied:
@djc Oops that has happened after the merge so... @FranciscoTGouveia Maybe you could fix it in a new PR? 🙏

FranciscoTGouveia replied:
Yes, of course! Thank you for pointing this out! :)

FranciscoTGouveia replied:
Doesn't this get overshadowed by this [comment](https://github.com/rust-lang/rustup/pull/4388#discussion_r2187891633)?

djc replied:
Potentially! If so, feel free to ignore my comment.

FranciscoTGouveia replied:
I realize now that it doesn’t overshadow the previous point, I am sorry.
I have made that change in the follow-up [PR](https://github.com/rust-lang/rustup/pull/4406).

rami3l commented:
As @FranciscoTGouveia and I have agreed to address the bonus points (https://github.com/rust-lang/rustup/pull/4388#discussion_r2187891633, https://github.com/rust-lang/rustup/pull/4388#discussion_r2187892015) in a later commit, I think this PR is ready to go.

Nice work @FranciscoTGouveia !
